src/core/cluster.py

```diff
@@ -143,20 +166,59 @@ def client_port(self) -> int:
                 2181 if TLS is not enabled
                 2182 if TLS is enabled
         """
-        if self.cluster.tls:
-            return SECURE_CLIENT_PORT
+        port = SECURE_CLIENT_PORT if self.cluster.tls else CLIENT_PORT
 
-        return CLIENT_PORT
+        return port
 
     @property
-    def endpoints(self) -> list[str]:
-        """The connection uris for all started ZooKeeper units.
-
-        Returns:
-            List of unit addresses
+    @retry(
+        wait=wait_fixed(5),
+        stop=stop_after_attempt(3),
+        retry=retry_if_exception_cause_type(LightKubeApiError),
+        reraise=True,
+    )
+    def endpoints_external(self) -> str:
+        """Comma-separated string of connection uris for all started ZooKeeper unit, for external access.
+
+        K8s only.
         """
-        return sorted(
-            [server.host if self.substrate == "k8s" else server.ip for server in self.servers]
+        auth = "plain" if not self.cluster.tls else "tls"
+        if self.config.expose_external is ExposeExternal.NODEPORT:
+            # We might have several of them if we run on multiple k8s nodes
+            return ",".join(
+                sorted(
+                    {
+                        f"{server.node_ip}:{self.unit_server.k8s.get_nodeport(auth)}"
+                        for server in self.servers
+                    }
+                )
+            )
+        else:
+            # ExposeExternal.LOADBALANCER only remaining possibility
```

Comment:
Done in d54223715a61ab3e75dda649d682af1c90cf33bc.

I really wanted to use [`typing.assert_never`](https://docs.python.org/3/library/typing.html#typing.assert_never) so that we would have both a runtime error and a static type check one, but I could not figure out a way to signify that `ExposeExternal.FALSE` is not possible at this point in the execution without adding a messy check. Better luck next time